### PR TITLE
Support Cairo 0.10.1

### LINKED ISSUE #497: Support Cairo 0.10.1
- [x] add `__validate_deploy__` to `IAccount` and presets
- [ ] tests
- [ ] docs

https://github.com/starkware-libs/cairo-lang/releases/tag/v0.10.1

### PR DESCRIPTION
**Pull request by martriay**
This PR:
- depends on: https://github.com/OpenZeppelin/nile/issues/267
- fixes #497, #508
- proposes a new format for function signature styling, with the goal of streamlining code editing as well as legibility
- abstracts common signer logic into a `BaseSigner` class
- leaves `__validate_deploy__` out of the `IAccount` interface since it's not callable once deployed
- adds docs for counterfactual deployments

### CODE REVIEW

**Review by ericnordelo — COMMENTED** [review]
Looking good! I will re-review and approve after publishing (it is drafted).

File: tox.ini
Comment (martriay) [inline tox.ini:23]: waiting for https://github.com/OpenZeppelin/nile/pull/268

File: tests/signers.py
Comment (ericnordelo) [inline tests/signers.py:10]:
```suggestion
from nile.common import get_contract_class, get_class_hash
```

File: tests/signers.py
Comment (ericnordelo) [inline tests/signers.py:74]:
```suggestion
        class_hash = get_class_hash(contract_name)
```

File: tests/signers.py
Comment (ericnordelo) [inline tests/signers.py:179]:
```suggestion
        self.class_hash = get_class_hash("Account")
```

File: tests/signers.py
Comment (ericnordelo) [inline tests/signers.py:197]:
```suggestion
        self.class_hash = get_class_hash("EthAccount")
```

File: tests/signers.py
Comment (ericnordelo) [inline tests/signers.py:79]:
```suggestion
            calldata=[class_hash],
```

**Review by ericnordelo — COMMENTED** [review]
Looking good @martriay! Just pointing out the refactor coming from `get_class_hash` and I think is good to go!

File: src/openzeppelin/account/IAccount.cairo
Comment (andrew-fleming) [inline src/openzeppelin/account/IAccount.cairo:18]:
```suggestion
    func isValidSignature(
            hash: felt,
            signature_len: felt,
            signature: felt*
        ) -> (isValid: felt) {
```
No more double indents for implicits?

File: tests/signers.py
Comment (andrew-fleming) [inline tests/signers.py:1]:
```suggestion
```
We don't need this, right?

File: tests/signers.py
Comment (andrew-fleming) [inline tests/signers.py:15]: 👍

**Review by andrew-fleming — COMMENTED** [review]
Very good PR, @martriay! I left a few comments and questions :)

File: src/openzeppelin/account/IAccount.cairo
Comment (martriay) [reply src/openzeppelin/account/IAccount.cairo:18]: we had that before? these are explicit though. i lean towards single indent but no strong opinions

Comment (andrew-fleming) [reply src/openzeppelin/account/IAccount.cairo:18]: Whoops! Wrong example, but yeah. Check the pre-cairo v0.10.0 releases:
https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.3.2/src/openzeppelin/account/presets/Account.cairo#L16-L24

I think the double indent is a bit easier to read, but either way works

**Comment by martriay** [comment]
PR should be ready! i think coverage/CI is blocked by https://github.com/ericnordelo/nile-coverage/pull/10

File: tests/signers.py
Comment (andrew-fleming) [inline tests/signers.py:57]:
```suggestion
```

**Review by andrew-fleming — COMMENTED** [review]
Almost there! I left a few super small comments. The docs look great also. You think we should give a brief mention to `declare_class` and `deploy_account` in `MockSigner utility`?

File: tox.ini
Comment (ericnordelo) [inline tox.ini]: Why -n 1?

Comment (martriay) [reply tox.ini]: heh it lasted just one second before i rolled it back. for some reason `auto` crashes on my machine

File: docs/modules/ROOT/pages/accounts.adoc
Comment (andrew-fleming) [inline docs/modules/ROOT/pages/accounts.adoc:248]:
```suggestion
* `deploy_account(state, calldata, salt=0, nonce=0, max_fee=0)`: returns a future of a counterfactual deployment.
```

**Review by andrew-fleming — APPROVED** [review]
I left a tiny suggestion; otherwise, this looks good to me!

File: docs/modules/ROOT/pages/accounts.adoc
Comment (ericnordelo) [inline docs/modules/ROOT/pages/accounts.adoc:147]:
```suggestion
Since `address` will ultimately depend on the `class_hash` and `calldata`, it's safe for the protocol to validate the signature and spend the funds on that address.
```

**Review by ericnordelo — APPROVED** [review]
Looking good to go! Just left a comment for a typo.

**Pull request merged**
